specs/merge/beacon-chain.md

```diff
@@ -0,0 +1,257 @@
+    epoch = get_current_epoch(state)
+    return xor(get_randao_mix(state, epoch), hash(randao_reveal))
+```
+
+#### `compute_time_at_slot` 
+
+```python
+def compute_time_at_slot(state: BeaconState, slot: Slot) -> uint64:
+    return uint64(state.genesis_time + slot * SECONDS_PER_SLOT)
+```
+
+### Beacon state accessors
+
+#### `get_recent_beacon_block_roots`
+
+```python
+def get_recent_beacon_block_roots(state: BeaconState, qty: uint64) -> Sequence[Bytes32]:
+    return [get_block_root_at_slot(state.slot - i) if GENESIS_SLOT + i < state.slot else Bytes32() for i in reversed(range(1, qty + 1))]
+```
+
+#### `get_evm_beacon_block_roots`
+
+```python
+def get_evm_beacon_block_roots(state: BeaconState) -> Sequence[Bytes32]:
+    num_block_roots = min(EVM_BLOCK_ROOTS_SIZE, SLOTS_PER_HISTORICAL_ROOT)
+    return get_recent_beacon_block_roots(state, num_block_roots)
+```
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)  # [Modified in The Merge]
+    process_operations(state, block.body)
+    process_application_payload(state, block.body)  # [New in The Merge]
+```
+
+#### Modified `process_eth1_data`
+
+*Note*: The function `process_eth1_data` is modified to update `state.eth1_data` with `eth1_data` of each block.
+
+```python
+def process_eth1_data(state: BeaconState, body: BeaconBlockBody) -> None:
+    state.eth1_data = body.eth1_data
+```
+
+#### Application payload processing
+
+##### `BeaconChainData`
+
+*Note*: `BeaconChainData` contains beacon state data that is used by the application state transition function.
+
+```python
+class BeaconChainData(Container):
+    slot: Slot
+    randao_mix: Bytes32
+    timestamp: uint64
+    recent_block_roots: Sequence[Bytes32]
+```
+
+##### `get_application_state`
+
+*Note*: `ApplicationState` class is an abstract class representing ethereum application state.
+
+Let `get_application_state(application_state_root: Bytes32) -> ApplicationState`  be the function that given the root hash returns a copy of ethereum application state. 
+The body of the function is implementation dependant.
+
+##### `application_state_transition`
+
+Let `application_state_transition(application_state: ApplicationState, beacon_chain_data: BeaconChainData, application_payload: ApplicationPayload) -> None` be the transition function of ethereum application state. 
+The body of the function is implementation dependant.
+
+*Note*: `application_state_transition` must throw `AssertionError` if either the transition itself or one of the post-transition verifications has failed.
+
+##### `process_application_payload`
+
+```python
+def process_application_payload(state: BeaconState, body: BeaconBlockBody) -> None:
+    """
+    Note: This function is designed to be able to be run in parallel with 
```

Comment:
It's worth noting that the other `process_block` functions in eth2 take ~2ms on average hardware - pragmatically there's not _that_ much to be gained from parallelization here - I would guess this falls in the category of being a "nice touch" - it makes testing and reasoning easier for example when functions are "pure" in the order-independent sense.

In general, making the eth1 state transition synchronous here leaves smaller margins for attestation production, an issue we're already seeing during finalization today, when the beacon chain is doing "nothing" - it would be good to define behaviors / describe more in the validator guide what the client should do when eth1 is momentarily unable to produce a block on time (for whatever reason, hickup in the client, hardware etc), what the _next_ proposer should do and other failure scenarios, as well as investigate / simulate tolerances for failures and delays.

Reply:
To clarify, the above comment is not taking into account future eth2 data requirements of course - the 2ms time is from current load which is mainly CPU-bound - with more data and shards, these timings may change dramatically, but at that point eth1 and eth2 might end up competing for the same (scarce) resource, disk, which also is hard to parallelize.

Reply:
Concerns you have listed above are important and addressing them is a part of readiness criteria for consensus upgrade. Transaction throughput (the size of the application block) is one of the evaluation criteria of the chain after the upgrade, attesters are in a heavy dependency on this property and time restriction that we have throughout a slot turns into restriction on the block size. The other potential approach to pushing the limits is [delayed execution](https://ethresear.ch/t/commit-to-pre-state-instead-of-post-state-on-the-executable-beacon-chain/8802) which comes with new complexity related to verification of gas limit per block.

Proposers may follow the _pending block_ technique in a way it is currently implemented by Ethereum clients which means that application block is always ready in advance and proposer just needs to pick it up.

The other side of the problem of synchronous interaction between consensus and application chains is resource exhaustion attack vector, its analysis and mitigation.

All the above is out of the scope of this spec but definitely is on the critical path to the upgrade.

Reply:
One thing to note is that the primary attestation issue today is seen on epoch boundaries where the epoch transition  must be processed before the block is created *and* before the block is gossiped at each step (to get the correct proposer shuffling to validate the signature).

My estimation is that this is causing late receiving of blocks across the network and thus the incorrect/late attestations. This epoch (shuffling) processing problem is *independent* of computation you do after block is gossiped/forwarded, and the epoch processing issue is a major target for optimizations.

If the epoch boundary block can be gossiped quickly (same speed as non-epoch boundary blocks), then there should be ample time (3s+) to compute the block's payload to attest. Eth1 blocks today are processed in 200ms so as long as you don't do that processing as a gossip validation, I don't see this compounding our current issue.

Reply:
Note! another parallelization design goal is that it ensures the application layer could be pulled out into a shard. By reducing the same-slot-processing dependencies between beacon and application-layer, we keep future designs open